### ISSUE TITLE
Arturia Minilab Mk2 - Relative #1 vs #2 oddity

### ISSUE THREAD

**Issue by TommyC81** [helgoboss/realearn#620]
I'm using the Minilab mk2 controller preset downloaded via ReaPack. In the controller preset all rotary encoders are defined as Relative 2.

However, as per the tested controller guide, I've configured all rotaries as Relative 1 on the Minilab mk2 itself.

Setting up my own preset (main compartment), virtual encoders are not automatically detected, they do however work fine when I manually select them (I can do input as expected).

If I change the rotary configuration in the Minilab mk2 controller preset to Relative 1 (instead of 2), then the main preset will correctly automatically detect the virtual encoders, BUT the rotary input will not work correctly.

If I instead change the configuration on the Minilab mk2 hardware itself so the rotaries are Relative 2, and use the default Minilab mk2 controller preset (rotaries configured as Relative 2), then the virtual encoders are detected correctly, BUT the rotary input won't work correctly.

To summarize: Using the default Minilab mk2 preset, virtual encoders are not automatically detected, but input works fine when virtual encoder is manually selected. By changing either the controller preset rotaries to Relative 1 OR changing Minilab mk2 rotaries to Relative 2, the virtual encoders will be automatically detected but will not work correctly.

Something is going wrong, and I'm not sure what it is. I'm using the latest pre-release version as of today.

Are controller presets also versioned, potentially causing this?

**Comment by TommyC81** [comment]
Just a quick update to say that I tested this fresh again today, and can confirm this odd behavior continues.

Pre-text:
All encoders on the hardware have been configured as "Relative #1".

To re-create:
1. Create fresh Reaper project (Reaper 6.63)
2. Add Instrument Track with ReaLearn instance (2.13.0-pre11)
3. Select "Arturia MiniLab mkII" in both Control input and Feedback output.
4. In controller compartment, select "Arturia MiniLab mkII" (v1.1.1 according to ReaPack).
5. Notice that Enc 1 (Channel 1, CC 112) is defined as "Encoder 2" (Relative type 2), and "Virtual Multi 1".
6. In the Controller compartment, select filter by source and turn encoder 1 on the hardware. Get result "No mapping matching...".
7. Go to Main compartment, and Add one mapping.
8. For the new mapping, click "Learn source", and turn encoder 1 on the hardware. Get result raw midi data (Channel 1, CC 112) and not "Virtual Multi 1". Note: Indicated as "Encoder (Relative type 2)".
9. Go back to Controller compartment, edit Encoder 1 and change it to "Encoder (Relative type 1)".
10. In the controller compartment, select filter by source and turn encoder 1 on the hardware. Get the expected mapping shown.
11. Go back to Main compartment, click "Learn source" on the only mapping, and turn encoder 1 on the hardware. Get expected result "Virtual Multi 1 (Enc 1)".
12. Map input from "Virtual Multi 1 (Enc 1)" to anything and note that it does not work as expected (it's seemingly the wrong encoder type).

**Comment by helgoboss** [comment]
Could be that I have a mismatch between the "Tested controllers" docs and the preset. And yes, from time to time I update new versions of controller presets. Major new version changes indicate a backward-incompatible change (which doesn't affect existing projects though if you don't explicitly reload the preset). I'll have a look. Thanks.

**Comment by TommyC81** [comment]
It's an odd one, because ReaLearn won't interpret the rotary input
correctly if I set the rotary type to 1 in ReaLearn, despite this being the
setting in the hardware.

It is also weird that ReaLearn can only correctly filter and connect rotary
input to the correct virtual controller if I configure the rotary as type 1
in ReaLearn (despite ReaLearn only being able to interpret type 2
correctly).

I tried a few different things to figure out what's going on, but it's all
very contradicting at the moment:

* Hardware configured for rotary type 1 output.
* ReaLearn can only interpret rotary input correctly if configured as type
2 in ReaLearn.
* ReaLearn can only automatically filter and identify virtual rotary if
configured as type 1 in ReaLearn (but can then not interpret the input
correctly).

So, it is working with the rotaries configured as type 2 in ReaLearn
(contrary to the type 1 configuration on the hardware), but then other
features (filtering, identify virtual rotary) don't work.

**Comment by helgoboss** [comment]
What's called "Type 1" in the hardware usually doesn't correspond to what's called "Type 1" in ReaLearn. It's totally unrelated. There's no standard whatsoever how to call these relative modes. If choosing "type 2" in ReaLearn works for you, then do it. Detecting the encoder type is a heuristic (just a guess), it doesn't have to be correct. The best guess is usually achieved if you rotate clock-wise while learning.

About the rest, I will have a look.

**Comment by helgoboss** [comment]
Mmh, there seems to be something wrong with processing relative messages, specifically those from the MiniLab. That's weird.

**Comment by TommyC81** [comment]
> What's called "Type 1" in the hardware usually doesn't correspond to what's called "Type 1" in ReaLearn. It's totally unrelated. There's no standard whatsoever how to call these relative modes. If choosing "type 2" in ReaLearn works for you, then do it. Detecting the encoder type is a heuristic (just a guess), it doesn't have to be correct. The best guess is usually achieved if you rotate clock-wise while learning.

Thanks, that's great information!

Are the ReaLearn rotary types aligned with the ones that Reaper use (see image below)?

![image](https://user-images.githubusercontent.com/1904092/177209390-e435fba6-55e3-41c2-bd06-9d02509324fd.png)

> Mmh, there seems to be something wrong with processing relative messages, specifically those from the MiniLab. That's weird.

If there's anything I can test, please don't hesitate to let me know!

**Comment by helgoboss** [comment]
> On a separate note, I did run MidiView on my AKAI MPK Mini mk3; although the output is set as relative which for Akai should produce "1" for increment and "127" for decrement, it intermittently outputs 2 and 126 instead (seemingly 1-2% of the time). No obvious reason why.

That's probably encoder acceleration. Many devices emit higher increments/decrements when you turn the encoder faster. This is why "Speed Max" / "Step Size Max" exists in ReaLearn, which allows you to adjust the effect of higher increments/decrements than 1.
